### ISSUE TITLE
.backward() API pre-requisite changes to support higher order derivatives

### ISSUE THREAD

**Issue by davidgonmar** [tinygrad/tinygrad#6164]
```python
import torch

x = torch.tensor(2.0, requires_grad=True)
y = torch.tensor(3.0)

intermediate = x + y

res = 2 + intermediate
res.backward()

print(f"Gradient of loss with respect to x: {x.grad}")
print(f"Gradient of loss with respect to intermediate: {intermediate.grad}")
```
On PyTorch, x.grad will be what you expect, and intermediate.grad will be None. On Tinygrad, intermediate.grad will be actually populated after the backward. PyTorch does this since keeping additional grads on memory is obviously worse, it only keeps leaf tensor's grads (like model params, etc). Since Tinygrad is lazy (and backward is done at LazyBuffer level, not Tensor level, x.grad and intermediate.grad are 'independent'), this does not incur in additional mem usage for tinygrad. However, if we want to make Tinygrad autodiff complete (additional derivatives, means autodiff at tensor level, as mentioned in https://github.com/tinygrad/tinygrad/issues/5858), not deleting 'intermediate' grads could make memory usage go up in certain cases.
(keep in mind torch still allows to keep some intermediate grads with an additional argument, the default is only to keep leaf grads).

Is it fine to change the Tinygrad API to behave like PyTorch? This could be seen as 'worse" but it's the way to have the .backward PyTorch-like API + complete autodiff without additional memory usage.

Note: the only thing I am doubting is that weights, for example, are not 'leafs' in tinygrad (unless you detach them) (a leaf is a tensor with no inputs) (they are not leafs because they invoke 'sub methods' usually, so the weight itself comes from some inputs at tensor level, usually). That means it would be difficult to keep track of what grads does the user want. Jax solves this by letting the user specify the inputs to differentiate w.r.t., so it might be a good option to consider. Looking for opinions since this could potentially change the autodiff API depending on the option chosen.

**Comment by chenyuxyz** [comment]
higher order derivatives is supported with the new gradient api
